crates/bevy_ecs/src/query/access_iter.rs

```diff
@@ -0,0 +1,521 @@
+use crate::{
+    component::{ComponentId, Components},
+    query::QueryData,
+};
+
+/// The data storage type that is being accessed.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum EcsAccessType {
+    /// Accesses [`Component`](crate::prelude::Component) data
+    Component(EcsAccessLevel),
+    /// Accesses [`Resource`](crate::prelude::Resource) data
+    Resource(ResourceAccessLevel),
```

Comment:
We can probably ignore resource access for this.  We don't allow mutable access (#17116), so it will never conflict.  And it's never valid to access resources through `EntityMut::get_components_mut` anyway (#20315).